Drivers: flash: flash_shell.c: Requires device on destructive operations

mrjimenez opened the pull request:
This patch avoids the use of the default device on destructive operations like erase and write. Allowing it might have catastrophic results like erasing parts or the whole of the application itself.

mrjimenez committed:
drivers: flash: flash_shell.c: Use the SHELL_HELP macro

This patch uses the new SHELL_HELP() macro in this file.

Signed-off-by: Marcelo Roberto Jimenez <marcelo.jimenez@gmail.com>

de-nordic reviewed (changes requested):
Changes requested.

Migration guide, like here https://github.com/zephyrproject-rtos/zephyr/pull/103139/changes#diff-cfe638170581ecbf5be17d5db65f1c13536725aeaa582127a042c66d0ff55eef, would also be needed.

mrjimenez committed:
docs: migration-guide: Mention flash erase and write change

Mentions that `flash erase` and `flash write` now require explicit
device arguments.

Signed-off-by: Marcelo Roberto Jimenez <marcelo.jimenez@gmail.com>

mrjimenez committed:
drivers: flash: flash_shell.c: Fix parsing of "flash read"

The parsing of the number of bytes to read in the read command was being
done in hexadecimal, causing unexpected behavior.

Now the number can be interpreted in decimal or hexadecimal if prefixed
with 0x.

Signed-off-by: Marcelo Roberto Jimenez <marcelo.jimenez@gmail.com>

mrjimenez commented on drivers/flash/flash_shell.c:
```diff
@@ -98,13 +98,29 @@ static int cmd_erase(const struct shell *sh, size_t argc, char *argv[])
+	(void)strtoul(argv[1], &endptr, 16);
+	if (*endptr == '\0') {
+		/* First argument is a number, not a device name */
+		shell_error(sh, "Device name required for erase operations");
```
Fixed

mrjimenez commented on drivers/flash/flash_shell.c:
```diff
@@ -98,13 +98,29 @@ static int cmd_erase(const struct shell *sh, size_t argc, char *argv[])
 	if (result) {
 		return result;
 	}
 	if (argc > 2) {
```
up?

de-nordic replied:
0 is "erase", 1 is device, 2 is address, 3 is optional size.
Size below is argv[2], not size?

mrjimenez committed:
drivers: flash: flash_shell.c: Requires device on destructive ops

This patch avoids the use of the default device on destructive
operations like erase and write. Allowing it might have catastrophic
results like erasing parts or the whole of the application itself.

Signed-off-by: Marcelo Roberto Jimenez <marcelo.jimenez@gmail.com>

kartben reviewed (dismissed):
Shell commands usage help text need to be updated to reflect that device is required (drop square brackets)

kartben commented:
> > Shell commands usage help text need to be updated to reflect that device is required (drop square brackets)
> 
> But I believe they have been. The change is only for "flash erase" and "flash write". Look:
> 
> ```bash
> uart:~$ flash
> flash: wrong parameter count
> flash - Flash shell commands
> Subcommands:
>   copy              : Copy data from one flash device to another
>                       Usage: copy <src_device> <dst_device> <src_offset> <dst_offset>
>                              <size>
>   erase             : Erase pages on a flash device
>                       Usage: erase <device> <page address> [<size>]
>   read              : Read data from a flash device
>                       Usage: read [<device>] <address> [<byte count>]
>   test              : Test flash device
>                       Usage: test [<device>] <address> <size> <repeat count>
>   write             : Write data to a flash device
>                       Usage: write <device> <address> <dword> [<dword>...]
>   load              : Load data into a flash device
>                       Usage: load [<device>] <address> <size>
>   page_info         : Get information about a flash page
>                       Usage: page_info [<device>] <address>
>   partitions        : Get partitionsformation
>                       Usage: partitions 
>   read_test         : Read test on a flash device
>                       Usage: read_test [<device>] <address> <size> <repeat count>
>   write_test        : Write test on a flash device
>                       Usage: write_test [<device>] <address> <size> <repeat count>
>   erase_test        : Erase test on a flash device
>                       Usage: erase_test [<device>] <address> <size> <repeat count>
>   erase_write_test  : Erase and write test on a flash device
>                       Usage: erase_write_test [<device>] <address> <size> <repeat count>
> uart:~$ 
> ```


Yes, sorry! Was looking at the test commands, sorry

mrjimenez commented:
Hi Folks,

What is missing for this patch to go in?
